### PR TITLE
Remove IChecksummedObject type.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
This type existed only for testing purposes.  But there's ahn existing pattern we already have to allow customized test behavior here.  Having two mechanisms for this was super confusing and led me down a painful debugging hole in one of my branches.

### CODE REVIEW

File: src/Workspaces/CoreTestUtilities/Remote/TestSerializerService.cs [hunk 1]
```diff
@@ -62,20 +58,17 @@ public override void WriteMetadataReferenceTo(MetadataReference reference, Objec
...
-        }
+        protected override MetadataReference ReadMetadataReferenceFrom(ObjectReader reader, CancellationToken cancellationToken)
+            => reader.ReadBoolean()
+                ? s_wellKnownReferences[reader.ReadRequiredString()]
+                : base.ReadMetadataReferenceFrom(reader, cancellationToken);
+
+        protected override Checksum CreateChecksum(AnalyzerReference reference, CancellationToken cancellationToken)
+            => reference is TestGeneratorReference generatorReference
+                ? generatorReference.Checksum
+                : base.CreateChecksum(reference, cancellationToken);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CoreTestUtilities/Remote/TestSerializerService.cs:69]: this is new, and replaced the existing hijacking point that was using IChecksummedObject.

**Comment by CyrusNajmabadi** [comment]
@ToddGrun @dibarbet ptal.

File: src/Workspaces/CoreTestUtilities/TestGeneratorReference.cs [hunk 2]
```diff
@@ -13,10 +13,10 @@ namespace Roslyn.Test.Utilities
     /// A simple deriviation of <see cref="AnalyzerReference"/> that returns the source generator
     /// passed, for ease in unit tests.
     /// </summary>
-    public class TestGeneratorReference : AnalyzerReference, IChecksummedObject
+    public class TestGeneratorReference : AnalyzerReference
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CoreTestUtilities/TestGeneratorReference.cs:16]: the only impl of this interface.

File: src/Workspaces/Core/Portable/Serialization/SerializerService.cs [hunk 3]
```diff
@@ -74,11 +74,6 @@ public Checksum CreateChecksum(object value, CancellationToken cancellationToken
         {
             cancellationToken.ThrowIfCancellationRequested();
 
-            if (value is IChecksummedObject checksummedObject)
-            {
-                return checksummedObject.Checksum;
-            }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Serialization/SerializerService.cs:80]: existed only to hijack.  but we can already do that throuhg normal virtual methods.

File: src/Workspaces/Core/Portable/Remote/ISerializerService.cs [hunk 4]
```diff
@@ -11,9 +11,6 @@ namespace Microsoft.CodeAnalysis.Serialization;
 internal interface ISerializerService : IWorkspaceService
 {
     void Serialize(object value, ObjectWriter writer, CancellationToken cancellationToken);
-
-    void SerializeParseOptions(ParseOptions options, ObjectWriter writer);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Remote/ISerializerService.cs:15]: didn't need to be on the itnerface.  was never called externally.

**Review by dibarbet — APPROVED** [review]

**Pull request merged**